business/workloads.go

```diff
@@ -445,35 +449,53 @@ func (in *WorkloadService) BuildLogOptionsCriteria(container, duration, isProxy,
 	return opts, nil
 }
 
+// getParsedLogs only works correctly when opts has Duration *and* TailLines set.
 func (in *WorkloadService) getParsedLogs(namespace, name string, opts *LogOptions) (*PodLog, error) {
 	k8sOpts := opts.PodLogOptions
 	// the k8s API does not support "endTime/beforeTime". So for bounded time ranges we need to
 	// 1) discard the logs after sinceTime+duration
 	// 2) manually apply tailLines to the remaining logs
-	isBounded := opts.Duration != nil
 	tailLines := k8sOpts.TailLines
-	if isBounded {
-		k8sOpts.TailLines = nil
-	}
-
-	podLog, err := in.k8s.GetPodLogs(namespace, name, &k8sOpts)
+	k8sOpts.TailLines = nil
 
+	logsReader, err := in.k8s.StreamPodLogs(namespace, name, &k8sOpts)
 	if err != nil {
 		return nil, err
 	}
 
-	lines := strings.Split(podLog.Logs, "\n")
-	entries := make([]LogEntry, 0)
+	defer func() {
+		e := logsReader.Close()
+		if e != nil {
+			log.Errorf("Error when closing the connection streaming logs of a pod: %s", e.Error())
+		}
+	}()
+
+	bufferedReader := bufio.NewReader(logsReader)
+
+	// It is known in advance that "tailLines" is the maximum number of
```

Comment:
Hmmm, interesting thought.  I'm not sure.  For situations where there is heavy logging, even a short time window can result in a large number of log entries, and tailLines prevents unnecessary lines from being sent back to the client.

If we remove tailLines we could always perform streaming, right?  If we did that and added maxLines (where ML is maybe something easily consumable, like 500 or maybe 1000, and make it configurable?), it may be OK.  But with streaming you would get the first ML lines, not the tailing ML lines, right?  Wouldit make any sense to use the circular buffer of ML size, in conjunction with streaming?

Reply:
> If we remove tailLines we could always perform streaming, right?

Yes! :-)

> If we did that and added maxLines (where ML is maybe something easily consumable, like 500 or maybe 1000, and make it configurable?), it may be OK. But with streaming you would get the first ML lines, not the tailing ML lines, right?

Yes! :-(

About the _maxLines_ I think it depends on who are you thinking by "easily consumable". I opened yesterday #5201 which is a draft-PR for the front-end. Currently, hide/show is broken (I'm working on it and it is almost done). With the virtual scrolling, the browser can easily digest 250k lines of logs (which is half of the size in the tests mentioned in the main comment). However, with 250k lines of logs, clearly a user won't read/inspect it all, but I do think it is nice to have those large results because of the find/hide/regex feature.

If we do want to limit _maxLines_ to a few hundreds, I think find/hide filters will need to move to server side to keep those features valuable. I think the server would return up to _maxLines_ of already filtered log lines rather than stopping after processing _maxLines_ raw logs.

> Wouldit make any sense to use the circular buffer of ML size, in conjunction with streaming?

I think it is possible. I added "streaming" at two points:
1. From cluster API feeding raw logs to Kiali
2. From Kiali feeding the processed logs to the client/browser.

The (1) is happening always, regardless of chosen options and that would already be a good improvement. The (2), sadly, cannot happen when using the circular list (or I can't think a way to do it).

Reply:
> Currently, hide/show is broken (I'm working on it and it is almost done).

It works again on the PR :-)

Reply:
OK, I understand your point about allowing a larger number of log lines, because Show/Hide is a powerful mechanism that can reduce the set easily, and the browser can deal with a fairly large set.  I'm OK if we want to try your more radical change here, and remove `tailLines`, as part of this attempt to increase our scale.  It would remove the need for two different approaches in the backend.

I'm thinking though, that maybe instead of `maxLines` being a configured limit, that `maxLines` could replace `tailLines` as a client option.  This would give the user control.  The dropdown tooltip could change to indicate it's the *first* N lines.  The entries could be much bigger, like [100, 1000 (default), 10000,  All Lines]. or soemthing like that.

If a user does hit the `maxLines` limit, we may need to indicate it somewhere on-screen, or maybe generate an informational toast

Do you think "Last 1m" is still good for the smallest duration?  Or would it make sense to offer something even smaller by default for logs, like "Last 30s" (note that currently a custom time range has a 1m minimum but that can't easily be changed)?

Reply:
> I'm OK if we want to try your more radical change here, and remove `tailLines`, as part of this attempt to increase our scale. It would remove the need for two different approaches in the backend.
> 
> I'm thinking though, that maybe instead of `maxLines` being a configured limit, that `maxLines` could replace `tailLines` as a client option.

Good! Let's try it. I'll change this PR.

> Do you think "Last 1m" is still good for the smallest duration? Or would it make sense to offer something even smaller

After removing tailLines, I do think something smaller makes sense. I would even add 15 and 10 secs, which are the smaller auto-refresh intervals. Although, perhaps 30secs as default would be OK.

Reply:
OK, I did the changes.
The only change that I didn't do is adding the smaller durations. I understand that would affect all pages using the duration drop-down. So, I'm unsure about it.

I used a default of 3K max lines. Looks like at the default duration of 10m, the bookinfo apps reach the 1K limit easily. If this is not OK, I can change it to 1K.